# Capstone Project - Team 11

## Capstone assignment

[Link to capstnone project assignment - Pyladies Ams](https://github.com/pyladiesams/Bootcamp-Data-Analysis-beginner-apr-may2020/tree/master/Capstone)

##### Libraries

- Pandas
- Matplotlib
- Seaborn (optional)
- NumPy (optional)

##### Capstone Task


1. Find a dataset that has at least 2 CSV files (to merge/join/concat them later)
2. Come up with 5 questions that you want to answer while exploring the dataset
3. Create one GitHub repository for your team Capstone project
4. Perform EDA (Exploratoty Data Analysis) on your dataset with basic visualisations
5. Make a 5 min presentation about your team findings and be ready to present it online. Add presentation slides to your GitHub repo.
6. Celebrate and wait for your certificate (send your name and surname for the certificate in a direct message on Slack to Alyona Galyeva)


## Data preprocessing

Go directly to a [summary](#Summary) of preprocessing to copy the commands used.

In [1]:
# importing essential libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

data1 = pd.read_csv('Mobile_App_Store_Data/appleStore.csv')
data2 = pd.read_csv('Mobile_App_Store_Data/appleStore_description.csv')

# merge data
data = pd.merge(data1, data2, how='outer', on=['id', 'id'])

# clean
data = data.drop('Unnamed: 0', axis=1)
data = data.drop('track_name_y', axis=1)
data = data.drop('size_bytes_y', axis=1)

# show
data.head(3)

In [16]:
# loading data
data1 = pd.read_csv('Mobile_App_Store_Data/appleStore.csv')
data1.head()

Unnamed: 0         id                                         track_name  \
0           1  281656475                                    PAC-MAN Premium   
1           2  281796108                          Evernote - stay organized   
2           3  281940292    WeatherBug - Local Weather, Radar, Maps, Alerts   
3           4  282614216  eBay: Best App to Buy, Sell, Save! Online Shop...   
4           5  282935706                                              Bible   

   size_bytes currency  price  rating_count_tot  rating_count_ver  \
0   100788224      USD   3.99             21292                26   
1   158578688      USD   0.00            161065                26   
2   100524032      USD   0.00            188583              2822   
3   128512000      USD   0.00            262241               649   
4    92774400      USD   0.00            985920              5320   

   user_rating  user_rating_ver     ver cont_rating   prime_genre  \
0          4.0              4.5   6.3.5          4+         Games   
1          4.0              3.5   8.2.2          4+  Productivity   
2          3.5              4.5   5.0.0          4+       Weather   
3          4.0              4.5  5.10.0         12+      Shopping   
4          4.5              5.0   7.5.1          4+     Reference   

   sup_devices.num  ipadSc_urls.num  lang.num  vpp_lic  
0               38                5        10        1  
1               37                5        23        1  
2               37                5         3        1  
3               37                5         9        1  
4               37                5        45        1

In [3]:
data1.shape

(7197, 17)

In [4]:
data2 = pd.read_csv('Mobile_App_Store_Data/appleStore_description.csv')
data2.head()

id                                         track_name  size_bytes  \
0  281656475                                    PAC-MAN Premium   100788224   
1  281796108                          Evernote - stay organized   158578688   
2  281940292    WeatherBug - Local Weather, Radar, Maps, Alerts   100524032   
3  282614216  eBay: Best App to Buy, Sell, Save! Online Shop...   128512000   
4  282935706                                              Bible    92774400   

                                            app_desc  
0  SAVE 20%, now only $3.99 for a limited time!\r...  
1  Let Evernote change the way you organize your ...  
2  Download the most popular free weather app pow...  
3  The eBay app is the best way to find anything ...  
4  On more than 250 million devices around the wo...

In [5]:
data2.shape

(7197, 4)

In [6]:
# merging 2 dataframes
data = pd.merge(data1, data2, how='outer', on=['id', 'id'])

In [7]:
data.head(3)

Unnamed: 0         id                                     track_name_x  \
0           1  281656475                                  PAC-MAN Premium   
1           2  281796108                        Evernote - stay organized   
2           3  281940292  WeatherBug - Local Weather, Radar, Maps, Alerts   

   size_bytes_x currency  price  rating_count_tot  rating_count_ver  \
0     100788224      USD   3.99             21292                26   
1     158578688      USD   0.00            161065                26   
2     100524032      USD   0.00            188583              2822   

   user_rating  user_rating_ver    ver cont_rating   prime_genre  \
0          4.0              4.5  6.3.5          4+         Games   
1          4.0              3.5  8.2.2          4+  Productivity   
2          3.5              4.5  5.0.0          4+       Weather   

   sup_devices.num  ipadSc_urls.num  lang.num  vpp_lic  \
0               38                5        10        1   
1               37                5        23        1   
2               37                5         3        1   

                                      track_name_y  size_bytes_y  \
0                                  PAC-MAN Premium     100788224   
1                        Evernote - stay organized     158578688   
2  WeatherBug - Local Weather, Radar, Maps, Alerts     100524032   

                                            app_desc  
0  SAVE 20%, now only $3.99 for a limited time!\r...  
1  Let Evernote change the way you organize your ...  
2  Download the most popular free weather app pow...

In [8]:
data.shape

(7197, 20)

In [9]:
# cleaning data

data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7197 entries, 0 to 7196
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        7197 non-null   int64  
 1   id                7197 non-null   int64  
 2   track_name_x      7197 non-null   object 
 3   size_bytes_x      7197 non-null   int64  
 4   currency          7197 non-null   object 
 5   price             7197 non-null   float64
 6   rating_count_tot  7197 non-null   int64  
 7   rating_count_ver  7197 non-null   int64  
 8   user_rating       7197 non-null   float64
 9   user_rating_ver   7197 non-null   float64
 10  ver               7197 non-null   object 
 11  cont_rating       7197 non-null   object 
 12  prime_genre       7197 non-null   object 
 13  sup_devices.num   7197 non-null   int64  
 14  ipadSc_urls.num   7197 non-null   int64  
 15  lang.num          7197 non-null   int64  
 16  vpp_lic           7197 non-null   int64  


In [10]:
# drop columns, previously index from data2, duplicates - track_name_y, size_bytes_y
data = data.drop('Unnamed: 0', axis=1)
data.shape

(7197, 19)

In [11]:
def drop_duplicate_columns(data):
    """Drops columns with duplicate data - track_name_y, size_bytes_y, 
    only if all values are same for duplicate columns"""
    
    track_name_y, size_bytes_y = True, True
    
    for index, row in data.iterrows():
        if row['track_name_x'] != row['track_name_y']:
            track_name_y = False
    
    for index, row in data.iterrows():
        if row['size_bytes_x'] != row['size_bytes_y']:
            size_bytes_y = False

    if track_name_y:
        data = data.drop('track_name_y', axis=1)
    if size_bytes_y:
        data = data.drop('size_bytes_y', axis=1)
        
    return data

data = drop_duplicate_columns(data)
data.shape

(7197, 17)

In [12]:
# check NaNs
data.isnull().sum()

id                  0
track_name_x        0
size_bytes_x        0
currency            0
price               0
rating_count_tot    0
rating_count_ver    0
user_rating         0
user_rating_ver     0
ver                 0
cont_rating         0
prime_genre         0
sup_devices.num     0
ipadSc_urls.num     0
lang.num            0
vpp_lic             0
app_desc            0
dtype: int64

In [13]:
# check if the currency is same
data.currency.isin({'currency':['USD']}).count()

7197

In [14]:
# cleaned data
data.head(3)

id                                     track_name_x  size_bytes_x  \
0  281656475                                  PAC-MAN Premium     100788224   
1  281796108                        Evernote - stay organized     158578688   
2  281940292  WeatherBug - Local Weather, Radar, Maps, Alerts     100524032   

  currency  price  rating_count_tot  rating_count_ver  user_rating  \
0      USD   3.99             21292                26          4.0   
1      USD   0.00            161065                26          4.0   
2      USD   0.00            188583              2822          3.5   

   user_rating_ver    ver cont_rating   prime_genre  sup_devices.num  \
0              4.5  6.3.5          4+         Games               38   
1              3.5  8.2.2          4+  Productivity               37   
2              4.5  5.0.0          4+       Weather               37   

   ipadSc_urls.num  lang.num  vpp_lic  \
0                5        10        1   
1                5        23        1   
2                5         3        1   

                                            app_desc  
0  SAVE 20%, now only $3.99 for a limited time!\r...  
1  Let Evernote change the way you organize your ...  
2  Download the most popular free weather app pow...

## Summary

Cleaned and merged data contain 7197 rows and 17 columns. There are no NaN values, all currencies are in USD. There are no duplicate columns.

To preprocess data, run the commands below:

In [15]:
import numpy as np
import pandas as pd

data1 = pd.read_csv('Mobile_App_Store_Data/appleStore.csv')
data2 = pd.read_csv('Mobile_App_Store_Data/appleStore_description.csv')

# merge data
data = pd.merge(data1, data2, how='outer', on=['id', 'id'])

# clean
data = data.drop('Unnamed: 0', axis=1)
data = data.drop('track_name_y', axis=1)
data = data.drop('size_bytes_y', axis=1)

# show
data.head(3)

id                                     track_name_x  size_bytes_x  \
0  281656475                                  PAC-MAN Premium     100788224   
1  281796108                        Evernote - stay organized     158578688   
2  281940292  WeatherBug - Local Weather, Radar, Maps, Alerts     100524032   

  currency  price  rating_count_tot  rating_count_ver  user_rating  \
0      USD   3.99             21292                26          4.0   
1      USD   0.00            161065                26          4.0   
2      USD   0.00            188583              2822          3.5   

   user_rating_ver    ver cont_rating   prime_genre  sup_devices.num  \
0              4.5  6.3.5          4+         Games               38   
1              3.5  8.2.2          4+  Productivity               37   
2              4.5  5.0.0          4+       Weather               37   

   ipadSc_urls.num  lang.num  vpp_lic  \
0                5        10        1   
1                5        23        1   
2                5         3        1   

                                            app_desc  
0  SAVE 20%, now only $3.99 for a limited time!\r...  
1  Let Evernote change the way you organize your ...  
2  Download the most popular free weather app pow...